File: src/converter.ts
export interface Cell {
  char: string;
  r: number;
  g: number;
  b: number;
}

export interface ConvertOptions {
  cols?: number;
  rows?: number;
  ramp?: string;
}

const DEFAULT_RAMP = ' .,:;+*?%S#@';
const DEFAULT_COLS = 80;
const DEFAULT_ROWS = 40;

export function convertFrame(
  pixels: Uint8Array,
  frameWidth: number,
  frameHeight: number,
  options?: ConvertOptions,
): Cell[][] {
  const cols = options?.cols ?? DEFAULT_COLS;
  const rows = options?.rows ?? DEFAULT_ROWS;
  const ramp = options?.ramp ?? DEFAULT_RAMP;

  if (ramp.length < 2 || ramp.length > 1024) {
    throw new RangeError(`ramp length must be between 2 and 1024, got ${ramp.length}`);
  }
  if (cols <= 0 || rows <= 0) {
    throw new RangeError(`cols and rows must be positive, got cols=${cols} rows=${rows}`);
  }
  if (frameWidth <= 0 || frameHeight <= 0) {
    throw new RangeError(
      `frameWidth and frameHeight must be positive, got ${frameWidth}x${frameHeight}`,
    );
  }
  if (pixels.length < frameWidth * frameHeight * 4) {
    throw new RangeError(
      `pixels buffer too small: expected at least ${frameWidth * frameHeight * 4} bytes`,
    );
  }

  const rampLen = ramp.length;
  const result: Cell[][] = [];

  for (let row = 0; row < rows; row++) {
    const rowCells: Cell[] = [];

    for (let col = 0; col < cols; col++) {
      const xStart = (col * frameWidth) / cols;
      const xEnd = ((col + 1) * frameWidth) / cols;
      const yStart = (row * frameHeight) / rows;
      const yEnd = ((row + 1) * frameHeight) / rows;

      const pxStart = Math.floor(xStart);
      const pxEnd = Math.ceil(xEnd);
      const pyStart = Math.floor(yStart);
      const pyEnd = Math.ceil(yEnd);

      let accR = 0, accG = 0, accB = 0, accW = 0;

      for (let py = pyStart; py < pyEnd; py++) {
        const wy = Math.min(py + 1, yEnd) - Math.max(py, yStart);
        if (wy <= 0) continue;

        for (let px = pxStart; px < pxEnd; px++) {
          const wx = Math.min(px + 1, xEnd) - Math.max(px, xStart);
          if (wx <= 0) continue;

          const w = wx * wy;
          const offset = (py * frameWidth + px) * 4;

          accR += pixels[offset] * w;
          accG += pixels[offset + 1] * w;
          accB += pixels[offset + 2] * w;
          accW += w;
        }
      }

      let r = 0, g = 0, b = 0;
      if (accW > 0) {
        r = Math.trunc(accR / accW);
        g = Math.trunc(accG / accW);
        b = Math.trunc(accB / accW);
      }

      // BT.601 luma
      const luma = 0.299 * r + 0.587 * g + 0.114 * b;
      let idx = Math.trunc((luma * (rampLen - 1)) / 255);
      if (idx >= rampLen) idx = rampLen - 1;

      rowCells.push({ char: ramp[idx], r, g, b });
    }

    result.push(rowCells);
  }

  return result;
}

File: src/converter.c
#include "converter.h"

#include <math.h>
#include <stdint.h>

int convert_frame(
    const uint8_t *pixels,
    int frame_w, int frame_h,
    int cols, int rows,
    const char *ramp, int ramp_len,
    Cell *out
) {
    if (!pixels || !ramp || !out)              return -1;
    if (ramp_len < 2 || ramp_len > 1024)      return -2;
    if (cols <= 0 || rows <= 0)               return -3;
    if (frame_w <= 0 || frame_h <= 0)         return -3;

    for (int row = 0; row < rows; row++) {
        for (int col = 0; col < cols; col++) {

            /* map this cell to its pixel region (float precision) */
            float x_start = (float)col       * frame_w / cols;
            float x_end   = (float)(col + 1) * frame_w / cols;
            float y_start = (float)row        * frame_h / rows;
            float y_end   = (float)(row + 1)  * frame_h / rows;

            int px_start = (int)x_start;
            int px_end   = (int)ceilf(x_end);   /* exclusive */
            int py_start = (int)y_start;
            int py_end   = (int)ceilf(y_end);   /* exclusive */

            float acc_r = 0.0f, acc_g = 0.0f, acc_b = 0.0f, acc_w = 0.0f;

            for (int py = py_start; py < py_end; py++) {
                float wy = fminf((float)(py + 1), y_end) - fmaxf((float)py, y_start);
                if (wy <= 0.0f) continue;

                for (int px = px_start; px < px_end; px++) {
                    float wx = fminf((float)(px + 1), x_end) - fmaxf((float)px, x_start);
                    if (wx <= 0.0f) continue;

                    float w = wx * wy;
                    const uint8_t *p = pixels + (py * frame_w + px) * 4;

                    acc_r += p[0] * w;
                    acc_g += p[1] * w;
                    acc_b += p[2] * w;
                    acc_w += w;
                }
            }

            uint8_t r = 0, g = 0, b = 0;
            if (acc_w > 0.0f) {
                r = (uint8_t)(acc_r / acc_w);
                g = (uint8_t)(acc_g / acc_w);
                b = (uint8_t)(acc_b / acc_w);
            }

            /* BT.601 luma */
            float luma = 0.299f * r + 0.587f * g + 0.114f * b;

            int idx = (int)(luma * (ramp_len - 1) / 255.0f);
            if (idx >= ramp_len) idx = ramp_len - 1;

            out[row * cols + col] = (Cell){ ramp[idx], r, g, b };
        }
    }

    return 0;
}
